### ISSUE TITLE
LinkedHashMap.keys behaves strangely after a map is modified

### ISSUE THREAD

**Issue by nex3** [dart-lang/sdk#38256]
The iterable returned by `LinkedHashMap.keys` behaves strangely in the presence of map modifications, and also inconsistently with `HashMap`:

```dart
import 'dart:collection';

void main() {
  print("LinkedHashMap:");
  var map = {"original": null};
  var keys = map.keys;
  map["added"] = null;
  print("  after a key is added: $keys");
  map.remove("original");
  print("  after a key is removed: $keys");

  print("HashMap:");
  map = HashMap.of({"original": null});
  keys = map.keys;
  map["added"] = null;
  print("  after a key is added: $keys");
  map.remove("original");
  print("  after a key is removed: $keys");
}
```

This prints:

```
LinkedHashMap:
  after a key is added: (original)
  after a key is removed: ()
HashMap:
  after a key is added: (original, added)
  after a key is removed: (added)
```

As you can see, `LinkedHashMap.keys` doesn't include a new key that's added after `.keys` was called, but it *does* omit a key that was removed. I would expect it either to represent the exact set of keys at the time it was called, to represent the exact set of keys at the time it was printed, or to throw an error if it was accessed after the map was modified.

The `HashMap.keys` iterable does the second option, representing the exact set of keys at the time it's printed.

**Comment by mraleph** [comment]
Implementation of `get keys` (and `get values`) does a strange thing of eagerly capturing some of the map's internal state - not sure why. I think `_data` and `_usedData` should not actually be captured until after `get iterator` is called. Seems like an easy fix.

```
  Iterable<K> get keys =>
      new _CompactIterable<K>(this, _data, _usedData, -2, 2);
```

**Comment by rakudrama** [comment]
@lrhn I believe you just fixed this issue with https://github.com/dart-lang/sdk/commit/d8fa6fe25522f45b7070278ad152abe4724cbdf2

**Comment by rakudrama** [comment]
Verified fixed by d8fa6fe25522f45b7070278ad152abe4724cbdf2
